### PR TITLE
M2P-412 Make phpunit green on Magento 2.4

### PR DESCRIPTION
**Pull request by dumega**
# Description
Make phpunit green on Magento 2.4

Fixes: https://boltpay.atlassian.net/browse/M2P-412

#changelog Make phpunit green on Magento 2.4

# Type of change

- [ ] Bug fix (change which fixes an issue)
- [x] New feature (change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?
Please validate that you have tested your change in at least one of the following areas:

- [ ] Successfully tested locally (or docker image)
- [ ] Successfully tested on a staging or sandbox server
- [ ] Successfully tested on a merchant's staging server

# For PR Reviewer 
- [ ] Reviewed unit tests to make sure we are using real components rather than mocks as much as possible?
- [ ] For any major change (observer, new Bolt feature, core Magento interaction) we must add a feature switch, did you verify this?

# Checklist:

- [ ] My code follows the style guidelines of this project.
- [ ] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] New and existing unit tests pass locally with my changes.
- [ ] I have created or modified unit tests to sufficiently cover my changes.
- [ ] I have added my Jira ticket link and provided a changelog message.

### CODE REVIEW

**Comment by vitaliyreznikov** [comment]
As discussed in PM, Ning is going to create separate PR with changes to the parent test class only, so it will be more convenient for discuss

File: Test/Unit/BoltTestCase.php [hunk 1]
```diff
@@ -3,13 +3,318 @@
...
+use PHPUnit\Framework\Constraint\IsType;
 
-class BoltTestCase extends TestCase
-{
-    protected function skipTestInUnitTestsFlow()
+if (!defined('PHP_CURRENT_VERSION')) {
+    define('PHP_CURRENT_VERSION', (float)phpversion());
+}
+
+if (PHP_CURRENT_VERSION < 7.2) {
```
Comment (danac-gs) [inline Test/Unit/BoltTestCase.php:16]: What's the logic behind this?
PHPUnit 9 requires PHP >= 7.3
Magento2.4 supports PHP 7.4 (eventually 7.3 but not tested or recommended)

**Comment by dumega** [comment]
@vitaliyreznikov @danac-gs Please see https://github.com/BoltApp/bolt-magento2/pull/1068/

**Comment by adentong** [comment]
Please rebase on top of latest master and uncomment the php7.4 magento2.4 unit test

File: Test/Unit/phpunit.xml [hunk 2]
```diff
@@ -48,8 +47,8 @@
     </listeners>
     <logging>
         <log type="coverage-text" target="php://stdout" showUncoveredFiles="false"/>
-        <coverage_html_placeholder>
+        <!--<coverage_html_placeholder>-->
             <log type="coverage-html" target="{{coverage_dir}}/test-reports/coverage" charset="UTF-8" yui="true" highlight="true"/>
-        </coverage_html_placeholder>
+        <!--</coverage_html_placeholder>-->
```
Comment (adentong) [inline Test/Unit/phpunit.xml:52]: instead of commenting out can we just remove?

**Review by adentong — DISMISSED** [review]
Really clever fix. Great job!

File: Test/Unit/phpunit.xml (hunk 2 above)
Comment (dumega) [reply Test/Unit/phpunit.xml]: Done

**Review by adentong — DISMISSED** [review]
LGTM

**Review by vitaliyreznikov — APPROVED** [review]

File: Test/Unit/TestHelper.php [hunk 3]
```diff
@@ -19,7 +19,9 @@ class TestHelper extends TestCase
      */
     public static function getReflectedClass($class)
     {
-        if (is_subclass_of($class, 'PHPUnit_Framework_MockObject_MockObject')) {
+        if (is_subclass_of($class, '\PHPUnit\Framework\MockObject\MockObject')) {
+            return new ReflectionClass(get_parent_class($class));
+        } else if (is_subclass_of($class, 'PHPUnit_Framework_MockObject_MockObject')) {
```
Comment (danac-gs) [inline Test/Unit/TestHelper.php:24]: as the result is the same for both conditions you can remove the `else if` clause and put both checks into `if`

File: Test/Unit/Helper/ApiTest.php [hunk 4]
```diff
@@ -394,13 +404,24 @@ public function sendRequest_withRegularRequest_returnsLocalizedException()
...
+                    'BOLT API REQUEST' => $request->getData()
+                ]
+            );
+        } else {
+            $boltApiRequest = new \Bolt\Boltpay\Test\Unit\ArraySubset(
+                [
+                    'BOLT API REQUEST' => $request->getData()
+                ]
+            );
+        }
```
Comment (danac-gs) [inline Test/Unit/Helper/ApiTest.php:421]: as this repeats you could extract it in a helper method and put it somewhere generally available

**Review by danac-gs — DISMISSED** [review]
Great stuff!

File: Test/Unit/Helper/ApiTest.php (hunk 4 above)
Comment (dumega) [reply Test/Unit/Helper/ApiTest.php]: Done

File: Test/Unit/TestHelper.php (hunk 3 above)
Comment (dumega) [reply Test/Unit/TestHelper.php]: Done

**Review by danac-gs — APPROVED** [review]

**Pull request merged**